File: source/clibCore/Interfaces/cpApi/CPContentClass.vb

Option Explicit On
Option Strict On

Imports Contensive.BaseClasses
Imports System.Runtime.InteropServices
Imports Contensive.Core.Models.Entity
Imports Contensive.Core.Controllers

Namespace Contensive.Core
    '
    ' comVisible to be activeScript compatible
    '
    <ComVisible(True)>
    <ComClass(CPContentClass.ClassId, CPContentClass.InterfaceId, CPContentClass.EventsId)>
    Public Class CPContentClass
        Inherits CPContentBaseClass
        Implements IDisposable
        '
#Region "COM GUIDs"
        Public Const ClassId As String = "D8D3D8F9-8459-46F7-B8AC-01B4DFAA4DB2"
        Public Const InterfaceId As String = "9B321DE5-D154-4EB1-B533-DBA2E5F2B5D2"
        Public Const EventsId As String = "6E068297-E09E-42C8-97B6-02DE591009DD"
#End Region
        '
        Private Property cp As CPClass
        Private Property cpCore As Contensive.Core.coreClass
        Protected Property disposed As Boolean = False
        '
        '====================================================================================================
        '
        Public Sub New(ByVal cpParent As CPClass)
            MyBase.New()
            cp = cpParent
            cpCore = cp.core
        End Sub
        '
        '====================================================================================================
        '
        Public Overrides Function GetCopy(ByVal CopyName As String, Optional ByVal DefaultContent As String = "") As String
            GetCopy = cpCore.html.html_GetContentCopy(CopyName, DefaultContent, cpCore.authContext.user.id, True, cpCore.authContext.isAuthenticated)
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function GetCopy(ByVal CopyName As String, ByVal DefaultContent As String, ByVal personalizationPeopleId As Integer) As String
            GetCopy = cpCore.html.html_GetContentCopy(CopyName, DefaultContent, personalizationPeopleId, True, cpCore.authContext.isAuthenticated)
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Sub SetCopy(ByVal CopyName As String, ByVal Content As String)
            Call cpCore.db.content_SetContentCopy(CopyName, Content)
        End Sub
        '
        '====================================================================================================
        '
        Public Overrides Function GetAddLink(ByVal ContentName As String, ByVal PresetNameValueList As String, ByVal AllowPaste As Boolean, ByVal IsEditing As Boolean) As String
            Return cpCore.html.main_GetRecordAddLink2(ContentName, PresetNameValueList, AllowPaste, IsEditing)
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function GetContentControlCriteria(ByVal ContentName As String) As String
            Return cpCore.metaData.content_getContentControlCriteria(ContentName)
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function GetFieldProperty(ByVal ContentName As String, ByVal FieldName As String, ByVal PropertyName As String) As String
            Return cpCore.metaData.GetContentFieldProperty(ContentName, FieldName, PropertyName)
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function GetID(ByVal ContentName As String) As Integer
            Return cpCore.metaData.getContentId(ContentName)
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function GetProperty(ByVal ContentName As String, ByVal PropertyName As String) As String
            Return cpCore.metaData.GetContentProperty(ContentName, PropertyName)
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function GetDataSource(ByVal ContentName As String) As String
            Return cpCore.metaData.getContentDataSource(ContentName)
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function GetEditLink(ByVal ContentName As String, ByVal RecordID As String, ByVal AllowCut As Boolean, ByVal RecordName As String, ByVal IsEditing As Boolean) As String
            If True Then
                Return cpCore.html.main_GetRecordEditLink2(ContentName, genericController.EncodeInteger(RecordID), AllowCut, RecordName, IsEditing)
            Else
                Return ""
            End If
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function GetLinkAliasByPageID(ByVal PageID As Integer, ByVal QueryStringSuffix As String, ByVal DefaultLink As String) As String
            Return docController.getLinkAlias(cpCore, PageID, QueryStringSuffix, DefaultLink)
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function GetPageLink(ByVal PageID As Integer, Optional ByVal QueryStringSuffix As String = "", Optional ByVal AllowLinkAlias As Boolean = True) As String
            Return cpCore.doc.getPageLink(PageID, QueryStringSuffix, AllowLinkAlias, False)
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function GetRecordID(ByVal ContentName As String, ByVal RecordName As String) As Integer
            Return cpCore.db.getRecordID(ContentName, RecordName)
        End Function
        '
        '====================================================================================================
        ''' <summary>
        ''' returns the matching record name if a match is found, otherwise blank. Does NOT validate the record.
        ''' </summary>
        ''' <param name="ContentName"></param>
        ''' <param name="RecordID"></param>
        ''' <returns></returns>
        Public Overrides Function GetRecordName(ByVal ContentName As String, ByVal RecordID As Integer) As String
            Return cpCore.db.getRecordName(ContentName, RecordID)
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function GetTable(ByVal ContentName As String) As String
            Return cpCore.metaData.getContentTablename(ContentName)
        End Function
        '
        '====================================================================================================
        '
        <Obsolete("Deprecated, template link is not supported", True)>
        Public Overrides Function GetTemplateLink(ByVal TemplateID As Integer) As String
            Return "" ' cpCore.doc.getTemplateLink(TemplateID)
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function IsField(ByVal ContentName As String, ByVal FieldName As String) As Boolean
            Return cpCore.metaData.IsContentFieldSupported(ContentName, FieldName)
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function IsLocked(ByVal ContentName As String, ByVal RecordID As String) As Boolean
            Return cpCore.workflow.isRecordLocked(ContentName, genericController.EncodeInteger(RecordID), 0)
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function IsChildContent(ByVal ChildContentID As String, ByVal ParentContentID As String) As Boolean
            Return cpCore.metaData.isWithinContent(genericController.EncodeInteger(ChildContentID), genericController.EncodeInteger(ParentContentID))
        End Function
        '
        '====================================================================================================
        '
        <Obsolete("workflow editing is deprecated", False)> Public Overrides Function IsWorkflow(ByVal ContentName As String) As Boolean
            '
            ' -- workflow no longer supported (but may come back)
            Return False
        End Function
        '
        '====================================================================================================
        '
        <Obsolete("workflow editing is deprecated", False)> Public Overrides Sub PublishEdit(ByVal ContentName As String, ByVal RecordID As Integer)
            ' Call cpCore.workflow.publishEdit(ContentName, RecordID, 0)
        End Sub
        '
        '====================================================================================================
        '
        <Obsolete("workflow editing is deprecated", False)> Public Overrides Sub SubmitEdit(ByVal ContentName As String, ByVal RecordID As Integer)
            'Call cpCore.workflow.submitEdit2(ContentName, RecordID, 0)
        End Sub
        '
        '====================================================================================================
        '
        <Obsolete("workflow editing is deprecated", False)> Public Overrides Sub AbortEdit(ByVal ContentName As String, ByVal RecordId As Integer)
            ' Call cpCore.workflow.abortEdit2(ContentName, RecordId, 0)
        End Sub
        '
        '====================================================================================================
        '
        <Obsolete("workflow editing is deprecated", False)> Public Overrides Sub ApproveEdit(ByVal ContentName As String, ByVal RecordId As Integer)
            'Call cpCore.workflow.approveEdit(ContentName, RecordId, 0)
        End Sub
        '
        '====================================================================================================
        '
        Public Overrides Function getLayout(ByVal layoutName As String) As String
            Dim result As String = ""
            Try
                Dim cs As New csController(cpCore)
                cs.open("layouts", "name=" & cp.Db.EncodeSQLText(layoutName), "id",, "layout")
                If cs.OK Then
                    result = cs.GetText("layout")
                End If
                Call cs.Close()
            Catch ex As Exception
                Call cpCore.handleException(ex) : Throw ' "Unexpected error in getLayout")
            End Try
            Return result
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function AddRecord(ByVal ContentName As String, ByVal recordName As String) As Integer
            Dim recordId As Integer = 0
            Try
                Dim cs As New csController(cpCore)
                If cs.Insert(ContentName) Then
                    cs.SetField("name", recordName)
                    recordId = cs.GetInteger("id")
                End If
                Call cs.Close()
            Catch ex As Exception
                Call cpCore.handleException(ex) : Throw ' "Unexpected error in AddRecord")
            End Try
            Return recordId
        End Function
        '
        '====================================================================================================
        '
        <Obsolete("Please use AddRecord( ContentName as string) ", True)>
        Public Overrides Function AddRecord(ContentName As Object) As Integer
            Return AddRecord(cp.Utils.encodeText(ContentName))
        End Function
        Public Overrides Function AddRecord(ContentName As String) As Integer
            Dim recordId As Integer = 0
            Try
                Dim cs As New csController(cpCore)
                If cs.Insert(ContentName) Then
                    recordId = cs.getInteger("id")
                End If
                Call cs.Close()
            Catch ex As Exception
                Call cpCore.handleException(ex) : Throw ' "Unexpected error in AddRecord")
            End Try
            Return recordId
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Sub Delete(ByVal ContentName As String, ByVal SQLCriteria As String)
            Call cpCore.db.deleteContentRecords(ContentName, SQLCriteria)
        End Sub
        '
        '====================================================================================================
        '
        Public Overrides Sub DeleteContent(ContentName As String)
            cpCore.metaData.deleteContent(ContentName)
        End Sub
        '
        '====================================================================================================
        '
        Public Overrides Function AddContentField(ContentName As String, FieldName As String, FieldType As Integer) As Integer
            Dim field As New CDefFieldModel()
            field.active = True
            field.adminOnly = False
            field.authorable = True
            field.blockAccess = False
            field.caption = FieldName
            field.contentId = cpCore.metaData.getContentId(ContentName)
            field.developerOnly = False
            field.editSortPriority = 9999
            field.editTabName = ""
            field.fieldTypeId = FieldType
            field.htmlContent = False
            field.indexColumn = 0
            field.indexSortDirection = 0
            field.indexSortOrder = 0
            field.indexWidth = ""
            field.installedByCollectionGuid = ""
            field.isBaseField = False
            field.lookupContentID = 0
            'field.lookupContentName = ""
            field.lookupList = ""
            field.manyToManyContentID = 0
            field.ManyToManyContentName(cpCore) = ""
            field.manyToManyRuleContentID = 0
            field.ManyToManyRuleContentName(cpCore) = ""
            field.ManyToManyRulePrimaryField = ""
            field.ManyToManyRuleSecondaryField = ""
            field.MemberSelectGroupID = 0
            field.MemberSelectGroupName(cpCore) = ""
            field.nameLc = FieldName.ToLower()
            field.Password = False
            field.ReadOnly = False
            field.RedirectContentID = 0
            field.RedirectContentName(cpCore) = ""
            field.RedirectID = ""
            field.RedirectPath = ""
            field.Required = False
            field.Scramble = False
            field.TextBuffered = False
            field.UniqueName = False
            Return cpCore.metaData.verifyCDefField_ReturnID(ContentName, field)
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function AddContent(ContentName As String) As Integer
            Return AddContent(ContentName, ContentName.Replace(" "c, "").Replace(" "c, ""), "default")
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function AddContent(ContentName As String, sqlTableName As String) As Integer
            Return AddContent(ContentName, sqlTableName, "default")
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function AddContent(ContentName As String, sqlTableName As String, dataSourceName As String) As Integer
            Return cpCore.metaData.createContent(True _
                , Models.Entity.dataSourceModel.create(cpCore, dataSourceName, New List(Of String)) _
                , sqlTableName _
                , ContentName _
                , False _
                , False _
                , True _
                , False _
                , "" _
                , "sort order" _
                , "name" _
                , False _
                , False _
                , False _
                , False _
                , False _
                , False _
                , "" _
                , 0 _
                , 0 _
                , 0 _
                , "" _
                , False _
                , "" _
                , True
                )
        End Function
        '
        '====================================================================================================
        '
        Public Overrides Function GetListLink(ContentName As String) As String
            Dim returnHtml As String = ""
            Try
                Dim adminUrl As String = cp.Site.GetText("adminUrl") & "?cid=" & cp.Content.GetID(ContentName)
                Dim encodedCaption As String = "List Records in " & cp.Utils.EncodeHTML(ContentName)

                returnHtml = returnHtml _
                    & "<a" _
                    & " class=""ccRecordEditLink"" " _
                    & " TabIndex=-1" _
                    & " href=""" & adminUrl & """" _
                    & "><img" _
                    & " src=""/ccLib/images/IconContentEdit.gif""" _
                    & " border=""0""" _
                    & " alt=""" & encodedCaption & """" _
                    & " title=""" & encodedCaption & """" _
                    & " align=""absmiddle""" _
                    & "></a>"
            Catch ex As Exception
                Call cpCore.handleException(ex) : Throw ' "Unexpected error in GetListLink")
            End Try
            Return returnHtml
        End Function
        '
        '====================================================================================================
        '
        Protected Overridable Overloads Sub Dispose(ByVal disposing As Boolean)
            If Not Me.disposed Then
                If disposing Then
                    '
                    ' call .dispose for managed objects
                    '
                    cpCore = Nothing
                    cp = Nothing
                End If
                '
                ' Add code here to release the unmanaged resource.
                '
            End If
            Me.disposed = True
        End Sub
        '
        '====================================================================================================
        '
#Region " IDisposable Support "
        ' Do not change or add Overridable to these methods.
        ' Put cleanup code in Dispose(ByVal disposing As Boolean).
        Public Overloads Sub Dispose() Implements IDisposable.Dispose
            Dispose(True)
            GC.SuppressFinalize(Me)
        End Sub
        Protected Overrides Sub Finalize()
            Dispose(False)
            MyBase.Finalize()
        End Sub
#End Region
    End Class
End Namespace

File: source/clibCoreCs/Interfaces/cpApi/CPContentClass.cs

using System;
using System.Reflection;
using System.Xml;
using System.Diagnostics;
using System.Linq;
using System.Text.RegularExpressions;
using System.Collections.Generic;
using System.Data;
using System.Data.SqlClient;
using Contensive.Core;
using Contensive.Core.Models.Entity;
using Contensive.Core.Controllers;
using static Contensive.Core.Controllers.genericController;
using static Contensive.Core.constants;
//
using Contensive.BaseClasses;
//
namespace Contensive.Core {
    public class CPContentClass : CPContentBaseClass, IDisposable {
        //
        #region COM GUIDs
        public const string ClassId = "D8D3D8F9-8459-46F7-B8AC-01B4DFAA4DB2";
        public const string InterfaceId = "9B321DE5-D154-4EB1-B533-DBA2E5F2B5D2";
        public const string EventsId = "6E068297-E09E-42C8-97B6-02DE591009DD";
        #endregion
        //
        private CPClass cp { get; set; }
        private Contensive.Core.coreClass cpCore { get; set; }
        protected bool disposed { get; set; } = false;
        //
        //====================================================================================================
        //
        public CPContentClass(CPClass cpParent) : base() {
            cp = cpParent;
            cpCore = cp.core;
        }
        //
        //====================================================================================================
        //
        public override string GetCopy(string CopyName, string DefaultContent = "") {
            return cpCore.html.html_GetContentCopy(CopyName, DefaultContent, cpCore.doc.authContext.user.id, true, cpCore.doc.authContext.isAuthenticated);
        }
        //
        //====================================================================================================
        //
        public override string GetCopy(string CopyName, string DefaultContent, int personalizationPeopleId) {
            return cpCore.html.html_GetContentCopy(CopyName, DefaultContent, personalizationPeopleId, true, cpCore.doc.authContext.isAuthenticated);
        }
        //
        //====================================================================================================
        //
        public override void SetCopy(string CopyName, string Content) {
            cpCore.db.SetContentCopy(CopyName, Content);
        }
        //
        //====================================================================================================
        //
        public override string GetAddLink(string ContentName, string PresetNameValueList, bool AllowPaste, bool IsEditing) {
            return cpCore.html.main_GetRecordAddLink2(ContentName, PresetNameValueList, AllowPaste, IsEditing);
        }
        //
        //====================================================================================================
        //
        public override string GetContentControlCriteria(string ContentName) {
            return Models.Complex.cdefModel.getContentControlCriteria(cpCore, ContentName);
        }
        //
        //====================================================================================================
        //
        public override string GetFieldProperty(string ContentName, string FieldName, string PropertyName) {
            return Models.Complex.cdefModel.GetContentFieldProperty(cpCore, ContentName, FieldName, PropertyName);
        }
        //
        //====================================================================================================
        //
        public override int GetID(string ContentName) {
            return Models.Complex.cdefModel.getContentId(cpCore, ContentName);
        }
        //
        //====================================================================================================
        //
        public override string GetProperty(string ContentName, string PropertyName) {
            return Models.Complex.cdefModel.GetContentProperty(cp.core, ContentName, PropertyName);
        }
        //
        //====================================================================================================
        //
        public override string GetDataSource(string ContentName) {
            return Models.Complex.cdefModel.getContentDataSource(cpCore, ContentName);
        }
        //
        //====================================================================================================
        //
        public override string GetEditLink(string ContentName, string RecordID, bool AllowCut, string RecordName, bool IsEditing) {
            if (true) {
                return cpCore.html.main_GetRecordEditLink2(ContentName, genericController.EncodeInteger(RecordID), AllowCut, RecordName, IsEditing);
            } else {
                return "";
            }
        }
        //
        //====================================================================================================
        //
        public override string GetLinkAliasByPageID(int PageID, string QueryStringSuffix, string DefaultLink) {
            return docController.getLinkAlias(cpCore, PageID, QueryStringSuffix, DefaultLink);
        }
        //
        //====================================================================================================
        //
        public override string GetPageLink(int PageID, string QueryStringSuffix = "", bool AllowLinkAlias = true) {
            return pageContentController.getPageLink(cpCore, PageID, QueryStringSuffix, AllowLinkAlias, false);
        }
        //
        //====================================================================================================
        //
        public override int GetRecordID(string ContentName, string RecordName) {
            return cpCore.db.getRecordID(ContentName, RecordName);
        }
        //
        //====================================================================================================
        /// <summary>
        /// returns the matching record name if a match is found, otherwise blank. Does NOT validate the record.
        /// </summary>
        /// <param name="ContentName"></param>
        /// <param name="RecordID"></param>
        /// <returns></returns>
        public override string GetRecordName(string ContentName, int RecordID) {
            return cpCore.db.getRecordName(ContentName, RecordID);
        }
        //
        //====================================================================================================
        //
        public override string GetTable(string ContentName) {
            return Models.Complex.cdefModel.getContentTablename(cpCore, ContentName);
        }
        //
        //====================================================================================================
        //
        [Obsolete("Deprecated, template link is not supported", true)]
        public override string GetTemplateLink(int TemplateID) {
            return ""; // cpCore.doc.getTemplateLink(TemplateID)
        }
        //
        //====================================================================================================
        //
        public override bool IsField(string ContentName, string FieldName) {
            return Models.Complex.cdefModel.isContentFieldSupported(cpCore, ContentName, FieldName);
        }
        //
        //====================================================================================================
        //
        public override bool IsLocked(string ContentName, string RecordID) {
            return cpCore.workflow.isRecordLocked(ContentName, genericController.EncodeInteger(RecordID), 0);
        }
        //
        //====================================================================================================
        //
        public override bool IsChildContent(string ChildContentID, string ParentContentID) {
            return Models.Complex.cdefModel.isWithinContent(cp.core, genericController.EncodeInteger(ChildContentID), genericController.EncodeInteger(ParentContentID));
        }
        //
        //====================================================================================================
        //
        [Obsolete("workflow editing is deprecated", false)]
        public override bool IsWorkflow(string ContentName) {
            //
            // -- workflow no longer supported (but may come back)
            return false;
        }
        //
        //====================================================================================================
        //
        [Obsolete("workflow editing is deprecated", false)]
        public override void PublishEdit(string ContentName, int RecordID) {
            // Call cpCore.workflow.publishEdit(ContentName, RecordID, 0)
        }
        //
        //====================================================================================================
        //
        [Obsolete("workflow editing is deprecated", false)]
        public override void SubmitEdit(string ContentName, int RecordID) {
            //Call cpCore.workflow.submitEdit2(ContentName, RecordID, 0)
        }
        //
        //====================================================================================================
        //
        [Obsolete("workflow editing is deprecated", false)]
        public override void AbortEdit(string ContentName, int RecordId) {
            // Call cpCore.workflow.abortEdit2(ContentName, RecordId, 0)
        }
        //
        //====================================================================================================
        //
        [Obsolete("workflow editing is deprecated", false)]
        public override void ApproveEdit(string ContentName, int RecordId) {
            //Call cpCore.workflow.approveEdit(ContentName, RecordId, 0)
        }
        //
        //====================================================================================================
        //
        public override string getLayout(string layoutName) {
            string result = "";
            try {
                csController cs = new csController(cpCore);
                cs.open("layouts", "name=" + cp.Db.EncodeSQLText(layoutName), "id", false, "layout");
                if (cs.OK()) {
                    result = cs.getText("layout");
                }
                cs.Close();
            } catch (Exception ex) {
                cpCore.handleException(ex); // "Unexpected error in getLayout")
                throw;
            }
            return result;
        }
        //
        //====================================================================================================
        //
        public override int AddRecord(string ContentName, string recordName) {
            int recordId = 0;
            try {
                csController cs = new csController(cpCore);
                if (cs.Insert(ContentName)) {
                    cs.setField("name", recordName);
                    recordId = cs.getInteger("id");
                }
                cs.Close();
            } catch (Exception ex) {
                cpCore.handleException(ex); // "Unexpected error in AddRecord")
                throw;
            }
            return recordId;
        }
        //
        //====================================================================================================
        //
        [Obsolete("Please use AddRecord( ContentName as string) ", true)]
        public override int AddRecord(object ContentName) {
            return AddRecord(cp.Utils.EncodeText(ContentName));
        }
        public override int AddRecord(string ContentName) {
            int recordId = 0;
            try {
                csController cs = new csController(cpCore);
                if (cs.Insert(ContentName)) {
                    recordId = cs.getInteger("id");
                }
                cs.Close();
            } catch (Exception ex) {
                cpCore.handleException(ex); // "Unexpected error in AddRecord")
                throw;
            }
            return recordId;
        }
        //
        //====================================================================================================
        //
        public override void Delete(string ContentName, string SQLCriteria) {
            cpCore.db.deleteContentRecords(ContentName, SQLCriteria);
        }
        //
        //====================================================================================================
        //
        public override void DeleteContent(string ContentName) {
            contentModel.delete(cpCore, Models.Complex.cdefModel.getContentId(cpCore, ContentName));
        }
        //
        //====================================================================================================
        //
        public override int AddContentField(string ContentName, string FieldName, int FieldType) {
            Models.Complex.CDefFieldModel field = new Models.Complex.CDefFieldModel();
            field.active = true;
            field.adminOnly = false;
            field.authorable = true;
            field.blockAccess = false;
            field.caption = FieldName;
            field.contentId = Models.Complex.cdefModel.getContentId(cpCore, ContentName);
            field.developerOnly = false;
            field.editSortPriority = 9999;
            field.editTabName = "";
            field.fieldTypeId = FieldType;
            field.htmlContent = false;
            field.indexColumn = 0;
            field.indexSortDirection = 0;
            field.indexSortOrder = 0;
            field.indexWidth = "";
            field.installedByCollectionGuid = "";
            field.isBaseField = false;
            field.lookupContentID = 0;
            //field.lookupContentName = ""
            field.lookupList = "";
            field.manyToManyContentID = 0;
            field.set_ManyToManyContentName(cpCore, "");
            field.manyToManyRuleContentID = 0;
            field.set_ManyToManyRuleContentName(cpCore, "");
            field.ManyToManyRulePrimaryField = "";
            field.ManyToManyRuleSecondaryField = "";
            field.MemberSelectGroupID = 0;
            field.set_MemberSelectGroupName(cpCore, "");
            field.nameLc = FieldName.ToLower();
            field.Password = false;
            field.ReadOnly = false;
            field.RedirectContentID = 0;
            field.set_RedirectContentName(cpCore, "");
            field.RedirectID = "";
            field.RedirectPath = "";
            field.Required = false;
            field.Scramble = false;
            field.TextBuffered = false;
            field.UniqueName = false;
            return Models.Complex.cdefModel.verifyCDefField_ReturnID(cpCore, ContentName, field);
        }
        //
        //====================================================================================================
        //
        public override int AddContent(string ContentName) {
            return AddContent(ContentName, ContentName.Replace(' '.ToString(), "").Replace(' '.ToString(), ""), "default");
        }
        //
        //====================================================================================================
        //
        public override int AddContent(string ContentName, string sqlTableName) {
            return AddContent(ContentName, sqlTableName, "default");
        }
        //
        //====================================================================================================
        //
        public override int AddContent(string ContentName, string sqlTableName, string dataSourceName) {
            var tmpList = new List<string> { };
            dataSourceModel dataSource = dataSourceModel.createByName(cpCore, dataSourceName, ref tmpList);
            return Models.Complex.cdefModel.addContent(cpCore, true, dataSource, sqlTableName, ContentName, false, false, true, false, "", "sort order", "name", false, false, false, false, false, false, "", 0, 0, 0, "", false, "", true);
        }
        //
        //====================================================================================================
        //
        public override string GetListLink(string ContentName) {
            string returnHtml = "";
            try {
                string adminUrl = cp.Site.GetText("adminUrl") + "?cid=" + cp.Content.GetID(ContentName);
                string encodedCaption = "List Records in " + cp.Utils.EncodeHTML(ContentName);

                returnHtml = returnHtml + "<a"
                    + " class=\"ccRecordEditLink\" "
                    + " TabIndex=-1"
                    + " href=\"" + adminUrl + "\""
                    + "><img"
                    + " src=\"/ccLib/images/IconContentEdit.gif\""
                    + " border=\"0\""
                    + " alt=\"" + encodedCaption + "\""
                    + " title=\"" + encodedCaption + "\""
                    + " align=\"absmiddle\""
                    + "></a>";
            } catch (Exception ex) {
                cpCore.handleException(ex); // "Unexpected error in GetListLink")
                throw;
            }
            return returnHtml;
        }
        //
        //====================================================================================================
        //
        protected virtual void Dispose(bool disposing) {
            if (!this.disposed) {
                if (disposing) {
                    //
                    // call .dispose for managed objects
                    //
                    cpCore = null;
                    cp = null;
                }
                //
                // Add code here to release the unmanaged resource.
                //
            }
            this.disposed = true;
        }
        //
        //====================================================================================================
        //
        #region  IDisposable Support 
        // Do not change or add Overridable to these methods.
        // Put cleanup code in Dispose(ByVal disposing As Boolean).
        public void Dispose() {
            Dispose(true);
            GC.SuppressFinalize(this);
        }
        ~CPContentClass() {
            Dispose(false);
            //INSTANT C# NOTE: The base class Finalize method is automatically called from the destructor:
            //base.Finalize();
        }
        #endregion
    }
}
